### PR TITLE
Netty 4 and reactive streams support

### PR DESCRIPTION
**Pull request by jroper**
* Upgraded to Netty 4
* Replaced iteratees implementations of IO with netty-reactive-streams and netty-reactive-streams-http

### CODE REVIEW

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaHttpServer.scala [hunk 1]
```diff
@@ -12,7 +12,7 @@ import java.net.InetSocketAddress
 import akka.util.ByteString
 import play.api._
 import play.api.http.DefaultHttpErrorHandler
-import play.api.libs.streams.Accumulator
+import play.api.libs.streams.{ MaterialiseOnDemandPublisher, Accumulator }
```
Comment (richdougherty) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaHttpServer.scala]: Use US spelling here?

File: framework/src/play-netty-server/src/main/resources/reference.conf [hunk 2]
```diff
@@ -31,22 +35,23 @@ play.server {
...
+        # SO_KEEPALIVE = false
 
-      # Set whether the TCP no delay flag is set
-      # child.tcpNoDelay = false
+        # Set whether the TCP no delay flag is set
+        # TCP_NODELAY = false
+      }
 
-      # Set the size of the backlog of TCP connections.  The default and exact meaning of this parameter is JDK specific.
-      # backlog = 100
```
Comment (richdougherty) [inline framework/src/play-netty-server/src/main/resources/reference.conf]: This stuff will need to go in the migration guides eventually. Do you want to make an issue so we remember?

File: framework/src/play-netty-server/src/main/scala/play/core/server/netty/NettyModelConversion.scala [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+          forwardedHeaderHandler,
+          headers,
+          connectionSecureProtocol = secureProtocol
+        )
+      }
+    }
+  }
+
+  /** Create an unparsed request header. Used when even Netty couldn't parse the request. */
+  def createUnparsedRequestHeader(requestId: Long, request: HttpRequest, _remoteAddress: InetSocketAddress, secureProtocol: Boolean) = {
```
Comment (richdougherty) [inline framework/src/play-netty-server/src/main/scala/play/core/server/netty/NettyModelConversion.scala]: We've discussed this before, but I'll nag you again… I'd prefer we didn't use `RequestHeader`s when there are parse errors, it feels like it's asking for security problems! :)

File: framework/src/play/src/main/scala/play/api/http/StandardValues.scala [hunk 4]
```diff
@@ -197,6 +197,7 @@ trait Status {
   val UNPROCESSABLE_ENTITY = 422
   val LOCKED = 423
   val FAILED_DEPENDENCY = 424
+  val UPGRADE_REQURIED = 426
```
Comment (richdougherty) [inline framework/src/play/src/main/scala/play/api/http/StandardValues.scala]: Typo, should be: `UPGRADE_REQUIRED`.

File: framework/src/play-netty-server/src/main/resources/reference.conf (hunk 2 above)
Comment (jroper) [inline framework/src/play-netty-server/src/main/resources/reference.conf]: Done #5016.

File: framework/src/play/src/main/scala/play/api/http/StandardValues.scala (hunk 4 above)
Comment (jroper) [inline framework/src/play/src/main/scala/play/api/http/StandardValues.scala]: Updated.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaHttpServer.scala (hunk 1 above)
Comment (jroper) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaHttpServer.scala]: Updated.

**Pull request merged**